src/Application/Console/Formatters/Json.php

```diff
@@ -0,0 +1,129 @@
+use Symfony\Component\Console\Output\OutputInterface;
+
+final class Json implements Formatter
+{
+    /** @var OutputInterface */
+    private $output;
+
+    public function __construct(InputInterface $input, OutputInterface $output)
+    {
+        $this->output = $output;
+    }
+
+    /**
+     * Format the result to the desired format.
+     *
+     * @param InsightCollection $insightCollection
+     * @param string            $dir
+     * @param array<string>     $metrics
+     *
+     * @throws Exception
+     */
+    public function format(
+        InsightCollection $insightCollection,
+        string $dir,
+        array $metrics
+    ): void {
+        $results = $insightCollection->results();
+
+        $data = [
+            "summary" => [
+                "code" => $results->getCodeQuality(),
+                "complexity" => $results->getComplexity(),
+                'architecture' => $results->getStructure(),
+                'style' => $results->getStyle(),
+                'security issues' => $results->getTotalSecurityIssues(),
+            ],
+        ];
+        $data += $this->issues($insightCollection, $metrics);
+
+        $json = json_encode($data);
+
+        if ($json === false) {
+            throw new InvalidArgumentException("Failed parsing result to JSON.");
+        }
+
+        $this->output->write($json);
+    }
+
+    /**
+     * Outputs the issues errors according to the format.
+     *
+     * @param InsightCollection $insightCollection
+     * @param array<string>     $metrics
+     *
+     * @return array<string, array<int, array<string, int|string>>|null>
+     */
+    public function issues(
+        InsightCollection $insightCollection,
+        array $metrics
+    ): array {
+        $previousCategory = null;
+
+        $data = [];
+        $current = null;
+
+        foreach ($metrics as $metricClass) {
+            /** @var Insight $insight */
+            foreach ($insightCollection->allFrom(new $metricClass()) as $insight) {
+                if (! $insight->hasIssue()) {
+                    continue;
+                }
+
+                $category = explode('\\', $metricClass);
+                $category = $category[count($category) - 2];
+
+                if ($previousCategory !== $category
+                    && $previousCategory !== null) {
+                    $data[$previousCategory] = $current;
+                    $current = [];
+                }
+
+                $previousCategory = $category;
+
+                if (! $insight instanceof HasDetails) {
+                    $current[] = [
+                        'title' => $insight->getTitle(),
+                        'insightClass' => $insight->getInsightClass(),
+                    ];
+                    continue;
+                }
+
+                foreach ($insight->getDetails() as $detail) {
+                    $information = explode(':', $detail);
+
+                    if (count($information) !== 3) {
```

Comment:
Okay I get it why **Type hint declaration** is not well displayed.
```
  src/Domain/Sniffs/SniffDecorator.php:32: Method \NunoMaduro\PhpInsights\Domain\Sniffs\SniffDecorator::register() does not have @return annotation for its traversable return value.
```

We have `SniffDecorator::register` exploded also, and information is greater than 3.

May we should check `count($information) < 3` and use array_shit for `file` and `line`, and implode to get rest of message ?

For `CyclomaticComplexityIsHigh`:
```
  src/Application/Console/Commands/AnalyseCommand.php: 17 cyclomatic complexity
```
May be we can assume to add only file and message if information count == 2 ?

Reply:
Yeah. We could do that.

We could also go with a third option and change getDetails to return a Detail object instead and let each formatter handle how they want to format the details?

Reply:
Yeah, it's would be better :wink:

Reply:
So I added a class for details now :) Hope I formatted everything correctly